flink/flink-1.14/src/main/java/io/tidb/bigdata/flink/connector/sink/TiDBDataStreamSinkProvider.java

```diff
@@ -0,0 +1,274 @@
+  private final TableSchema tableSchema;
+  private final TiDBSinkOptions sinkOptions;
+
+  public  TiDBDataStreamSinkProvider(String databaseName, String tableName,
+      ResolvedCatalogTable table, Context context, TiDBSinkOptions sinkOptions) {
+    this.databaseName = databaseName;
+    this.tableName = tableName;
+    this.table = table;
+    this.context = context;
+    this.properties = table.getOptions();
+    this.tableSchema = TableSchemaUtils.getPhysicalSchema(table.getSchema());
+    this.sinkOptions = sinkOptions;
+  }
+
+  private byte[] fakePrimaryKey() {
+    return (PRIMARY_KEY_PREFIX + UUID.randomUUID()).getBytes();
+  }
+
+  private DataStream<Row> deduplicate(DataStream<Row> tiRowDataStream, TiTableInfo tiTableInfo) {
+    List<TiIndexInfo> uniqueIndexes = SqlUtils.getUniqueIndexes(tiTableInfo,
+        sinkOptions.isIgnoreAutoincrementColumn());
+    if (uniqueIndexes.size() == 0) {
+      return tiRowDataStream;
+    }
+
+    for (TiIndexInfo uniqueIndex : uniqueIndexes) {
+      List<Integer> columnIndexes = uniqueIndex.getIndexColumns()
+          .stream()
+          .map(TiIndexColumn::getOffset)
+          .collect(Collectors.toList());
+      List<String> uniqueIndexColumnNames = uniqueIndex.getIndexColumns()
+          .stream()
+          .map(TiIndexColumn::getName)
+          .collect(Collectors.toList());
+
+      tiRowDataStream = tiRowDataStream.keyBy(new KeySelector<Row, List<Object>>() {
+        @Override
+        public List<Object> getKey(Row row) throws Exception {
+          return columnIndexes.stream()
+              .map(i -> row.get(i, null))
+              .collect(Collectors.toList());
+        }
+      }).process(
+          TiDBKeyedProcessFunctionFactory.createKeyedProcessFunction(
+              sinkOptions,
+              tiRowDataStream,
+              uniqueIndexColumnNames));
+    }
+    return tiRowDataStream;
+  }
+
+  private List<Long> createRowIdStarts(ClientSession session, int parallelism) {
+    // create row id start list
+    int step = sinkOptions.getRowIdAllocatorStep();
+    long start = session.createRowIdAllocator(databaseName, tableName, step * parallelism,
+        3).getStart();
+    final List<Long> rowIdStarts = LongStream.range(0, parallelism)
+        .boxed()
+        .map(i -> start + i * step)
+        .collect(Collectors.toList());
+    LOG.info("Create row id starts success, rowIdStarts = " + rowIdStarts);
+    return rowIdStarts;
+  }
+
+  private DataStreamSink<?> consumeDataStream(DataStream<RowData> dataStream,
+      ClientSession clientSession) {
+    final byte[] primaryKey;
+    final int parallelism = dataStream.getParallelism();
+    final TiTimestamp timestamp = clientSession.getSnapshotVersion();
+    final SinkTransaction sinkTransaction = sinkOptions.getSinkTransaction();
+    TiTableInfo tiTableInfo = clientSession.getTableMust(databaseName, tableName);
+
+    // check if tidbColumns match flinkColumns
```

Comment:
when will the mismatch occur?

Reply:
@humengyu2012 can you explain this

Reply:
If the columns do not match, we may need to deal with a lot of extra things. But using catalog will avoid this situation. We can delete this check if table factory is deprecated.